### PR TITLE
Revamp of unit tests

### PR DESCRIPTION
**Pull request by tfdahlin**
With changes to YouTube that break pytube, it will sometimes be necessary to generate new mock files for the test suite. This adds that functionality to the `helpers.py` file, and also creates unit testing for that new function. 

As part of this pull request, the previous mock file that references Gangnam Style has been replaced with a mock file that references YouTube Rewind 2019. Additionally, existing unit tests have been updated to reflect this new mock file. 

Finally, there is another fix to make another existing unit test os-agnostic.

### CODE REVIEW

**Comment by tfdahlin** [comment]
cc: @RONNCC

**Comment by tfdahlin** [comment]
Technically this functionality *sorta* exists in the cli, but lacks unit-tests, doesn't save the result to the mock folder, and makes use of timestamps. It looks like that functionality is meant more for debugging. This simplifies the process of creating these files and includes unit test coverage for the function.

**Comment by tfdahlin** [comment]
Looks like the travis CI is failing due to youtube rate limiting. I'll figure out a way to create a mock instead so we can avoid that.

**Comment by tfdahlin** [comment]
This one's a bit of a doozie.
Firstly, it gets rid of the unit tests against gangnam style, in favor of unit tests against YouTube Rewind 2019, as requested as a follow-up to https://github.com/nficano/pytube/pull/767. 

In doing this, I discovered that the mock json files were generated kind of weird, and were sort of a partial serialization of a YouTube object. At first, I tried to create a function that replicated this (without realizing one existed), but while writing unit tests to make sure that it worked correctly, I ran into problems with Travis CI. Notably, I ran into `HTTP 429` responses from YouTube, so I had to work out how to use mock to get around that. As a result of this, I discovered I needed to create a cache of different webpages to test against, so I ended up reworking some of the unit tests to use pure HTML responses from a mock file, instead of using a partial serialization. Travis CI should no longer have rate limiting issues while running the unit test that I wrote for my serialization function. 

I also included another minor fix to unit tests where it was using platform-specific separators, rather than platform-agnostic.

Finally, I also attempted to follow up on #743 and I believe the way that I refactored __main__ will fix this problem.

cc: @RONNCC

File: tests/test_helpers.py [hunk 1]
```diff
@@ -90,3 +95,79 @@ def test_setup_logger(logging):
...
+    # The file header includes time metadata, so *occasionally* a single
+    #  byte will be off at the very beginning. In theory, this difference
+    #  should only affect bytes 5-8 (or [4:8] because of zero-indexing),
+    #  but I've excluded the 10-byte metadata header altogether from the
+    #  check, just to be safe.
+    # Source: https://en.wikipedia.org/wiki/Gzip#File_format
+    assert gzip_data[10:] == full_content[10:]
+
+
+def test_uniqueify():
```
Comment (RONNCC) [inline tests/test_helpers.py:165]: We can use python sets for this -- we dont need a uniqueify function. 

In scientific/number crunching context there's also numpy.unique

**Review by RONNCC — CHANGES_REQUESTED** [review]

**Comment by RONNCC** [comment]
@tfdahlin  one nit -- but otherwise looks good

File: tests/test_helpers.py (hunk 1 above)
Comment (tfdahlin) [reply tests/test_helpers.py:165]: @RONNCC unfortunately, we *do* need to use the existing uniqueify function. While python sets can be used for this, the problem is that converting from list -> set -> list loses the ordering. 

In the case of processing playlists (the only location uniqueify is used), the ordering of the video list is important, so this uniqueify function serves a purpose that cannot be accomplished simply by using sets.

**Comment by tfdahlin** [comment]
see response in conversation

File: tests/test_helpers.py [hunk 2]
```diff
@@ -90,3 +95,79 @@ def test_setup_logger(logging):
...
+
+    file_handle = mock_open.return_value.__enter__.return_value
+
+    # For some reason, write gets called multiple times, so we have to
+    #  concatenate all the write calls to get the full data before we compare
+    #  it to the BytesIO object value.
+    full_content = b''
+    for call in file_handle.write.call_args_list:
+        args, kwargs = call
+        for arg in args:
```
Comment (RONNCC) [inline tests/test_helpers.py]: nit: less verbose? --  full_content = ''.join(args)

File: tests/test_helpers.py (hunk 1 above)
Comment (RONNCC) [reply tests/test_helpers.py:165]: good point -- can use ordered dict (python 3.5+ dicts are ordered by default also, that that's breaking API contract a bit) 

https://repl.it/@spamspam3/VoluminousSilverPrograms

File: tests/test_helpers.py [hunk 3]
```diff
@@ -90,3 +95,79 @@ def test_setup_logger(logging):
...
+
+
+def test_uniqueify():
+    non_unique_list = [1, 2, 3, 3, 4, 5]
+    expected = [1, 2, 3, 4, 5]
+    result = uniqueify(non_unique_list)
+    assert len(expected) == len(result)
+
+    for i in range(len(expected)):
+        assert expected[i] == result[i]
```
Comment (RONNCC) [inline tests/test_helpers.py]: can just use assert list1 == list2. It checks order, type, and index. 
though usually you would subclass unittest pkg for these hmm

**Comment by RONNCC** [comment]
@tfdahlin  responded

**Comment by tfdahlin** [comment]
Good point on those tests, changed according to your suggestions @RONNCC

**Comment by RONNCC** [comment]
this is a giant set of changes -- thank you @tfdahlin  😄

**Pull request merged**